core/GlobalState.cc

```diff
@@ -1412,6 +1418,72 @@ bool GlobalState::wasModified() const {
     return wasModified_;
 }
 
+bool GlobalState::isTypecheckingCanceled() const {
+    return lspEpochInvalidator->load() != currentlyProcessingLSPEpoch->load();
+}
+
+void GlobalState::startCommitEpoch(u4 epoch) {
+    absl::MutexLock lock(epochMutex.get());
+    currentlyProcessingLSPEpoch->store(epoch);
+    lspEpochInvalidator->store(epoch);
+    // These should always be different.
+    ENFORCE(epoch != lastCommittedLSPEpoch->load());
+}
+
+optional<pair<u4, u4>> GlobalState::getRunningSlowPath() const {
+    absl::MutexLock lock(epochMutex.get());
+    const u4 processing = currentlyProcessingLSPEpoch->load();
+    const u4 committed = lastCommittedLSPEpoch->load();
+    if (processing == committed) {
+        return nullopt;
+    }
+    return make_pair(committed, processing);
+}
+
+bool GlobalState::tryCancelSlowPath(u4 newEpoch) const {
+    absl::MutexLock lock(epochMutex.get());
+    const u4 processing = currentlyProcessingLSPEpoch->load();
+    ENFORCE(newEpoch != processing); // This would prevent a cancelation from happening.
+    const u4 committed = lastCommittedLSPEpoch->load();
+    // The second condition should never happen, but guard against it in production.
+    if (processing == committed || newEpoch == processing) {
+        return false;
+    }
+    // Cancel slow path by bumping invalidator.
+    lspEpochInvalidator->store(newEpoch);
+    return true;
+}
+
+bool GlobalState::tryCommitEpoch(u4 epoch, bool isCancelable, function<bool()> lambda) {
+    if (!isCancelable) {
+        return lambda();
+    }
+
+    // Should have called "startCommitEpoch" *before* this method.
+    ENFORCE(currentlyProcessingLSPEpoch->load() == epoch);
+    const bool canCommit = lambda();
```

Comment:
should this execute under the mutex?

also, in the example you have the `lambda` seems costly, can we run some cheap checks to see if it's too late to cachel?
e.g. if `processing != invalidator` we can't cancel, right?

Reply:
I'll add a comment. `lambda` runs typechecking, so it wouldn't make sense to hold the mutex  -- it would prevent cancelation from occurring.

Reply:
I didn't get your second comment, since I only ever use `processing != invalidator` to check if we can cancel, but now I realize that the `bool` return value from the lambda is confusing.

It actually doesn't need to return anything. I'll fix that.